libbeat/publisher/queue/spool/outbroker.go

```diff
@@ -0,0 +1,489 @@
+package spool
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/elastic/beats/libbeat/publisher"
+	"github.com/elastic/xbeats/txfile/pq"
+)
+
+type outBroker struct {
+	ctx   *spoolCtx
+	state func(*outBroker) bool
+
+	// internal API
+	sigFlushed chan uint
+	get        chan getRequest
+
+	// ack signaling
+	pendingACKs   chanList      // list of pending batches to be forwarded to the ackLoop
+	scheduledACKs chan chanList // shared channel for forwading batches to ackLoop
+	schedACKs     chan chanList // active ack forwarding channel, as used by broker (nil if pendingACKs is empty)
+
+	// queue state
+	queue     *pq.Queue
+	available uint // number of available events. getRequests are only accepted if available > 0
+	events    []publisher.Event
+	required  int
+	total     int
+	active    getRequest
+
+	// internal
+	timer *timer
+	dec   *decoder
+}
+
+type chanList struct {
+	head *ackChan
+	tail *ackChan
+}
+
+type ackChan struct {
+	next  *ackChan
+	ch    chan batchAckMsg
+	total int // total number of events to ACK with this batch
+}
+
+const (
+	// maximum number of events if getRequest size is <0
+	maxEvents = 2048
+)
+
+var ackChanPool = sync.Pool{
+	New: func() interface{} {
+		return &ackChan{
+			ch: make(chan batchAckMsg, 1),
+		}
+	},
+}
+
+var errRetry = errors.New("retry")
```

Comment:
I had to check where it was used to understand the context of this error, from what I see it is used to signal that we need to **retry to decode the event** when no **events were decoded**. Maybe `errNoEventDecoded` with a better message?

Reply:
It's named `errRetry`, as it's purpose is to mostly act as a signal to the consumer to retry (ignore error). Like, we didn't get an event for you, please retry. The retry signal is required to guarantee a consumer can be correctly closed asynchronously.